packager/packager.cc

```diff
@@ -1546,9 +1546,83 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
     return info;
 }
 
+void validateLayering(const core::Context &ctx, const Import &i) {
+    if (i.type == ImportType::Test) {
+        return;
+    }
+
+    const auto &packageDB = ctx.state.packageDB();
+    ENFORCE(packageDB.getPackageInfo(i.name.mangledName).exists())
+    ENFORCE(packageDB.getPackageForFile(ctx, ctx.file).exists())
+    auto &thisPkg = PackageInfoImpl::from(packageDB.getPackageForFile(ctx, ctx.file));
+    auto &otherPkg = PackageInfoImpl::from(packageDB.getPackageInfo(i.name.mangledName));
+
+    if (!thisPkg.strictDependenciesLevel.has_value() || !otherPkg.strictDependenciesLevel.has_value() ||
+        !thisPkg.layer.has_value() || !otherPkg.layer.has_value()) {
+        return;
+    }
+
+    if (thisPkg.strictDependenciesLevel.value().first == core::packages::StrictDependenciesLevel::False) {
+        return;
+    }
+
+    auto possibleLayers = packageDB.layers();
+    auto pkgLayer = thisPkg.layer.value().first;
+    auto otherPkgLayer = otherPkg.layer.value().first;
+    auto pkgIndex = packageDB.layerIndex(pkgLayer);
+    auto otherPkgIndex = packageDB.layerIndex(otherPkgLayer);
+
+    if (pkgIndex < otherPkgIndex) {
+        if (auto e = ctx.beginError(i.name.loc, core::errors::Packager::LayeringViolation)) {
+            e.setHeader("`{}` is at layer `{}`, so it can not import package `{}`, which is at layer `{}`",
+                        thisPkg.name.toString(ctx), pkgLayer.toString(ctx), otherPkg.name.toString(ctx),
+                        otherPkgLayer.toString(ctx));
+            e.addErrorLine(core::Loc(thisPkg.loc.file(), thisPkg.layer.value().second), "`{}`'s `{}` declared here",
+                           thisPkg.name.toString(ctx), "layer");
+            e.addErrorLine(core::Loc(otherPkg.loc.file(), otherPkg.layer.value().second), "`{}`'s `{}` declared here",
+                           otherPkg.name.toString(ctx), "layer");
+        }
+    }
+
+    if (otherPkg.strictDependenciesLevel.value().first == core::packages::StrictDependenciesLevel::False) {
+        if (auto e = ctx.beginError(i.name.loc, core::errors::Packager::LayeringViolation)) {
+            e.setHeader("All of this package's dependecies must be `{}` or higher", "layered");
+            e.addErrorLine(core::Loc(otherPkg.loc.file(), otherPkg.strictDependenciesLevel.value().second),
+                           "`{}`'s `{}` level declared here", otherPkg.name.toString(ctx), "strict_dependencies");
+        }
+    }
```

Comment:
Do we need an autocorrect for either of these errors to get to pairity with gen-packages or do these errors not have autocorrects in gen-packages?

Reply:
There's no autocorrect from what I can tell: http://go/neil-gen-packages-layer-error